[16.0][ADD] l10n_es_atc_dua

julianram opened the pull request:
Importaciones con DUA ATC

julianram committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

# This is the commit message #2:

[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

julianram commented:
PR #3667  problemas con runbot.

arielbarreiros96 commented:
@julianram following OCA guidelines, the PR name should be in this case [16.0][ADD] l10n_es_atc_dua, also consider to squash all commits into a single commit to make it easier to review. I've also seen that the Readme and the name on the manifest are in Spanish, please consider to translate it to English, it is a OCA guideline gut it will also help others to review and maintain your code 😉

etobella commented:
Puedes juntar todos los commits en uno solo?

etobella commented:
@julianram A ver, tras revisarmelo un poco me surgen bastantes dudas. Por que no partís de l10n_es_dua para ya tener la posición fiscal y el impuesto DUA EXento? Lo comento por que en caso de empresas con doble sede será un follón a la hora de gestionar los impuestos. De esta forma debería ser mucho más sencillo

julianram commented:
@etobella  Si partimos de l10n_es_dua obligamos a las empresas en Canarias que nunca gestionaran DUAS de importación en  la AEAT a tenerlo instalado , las empresas con establecimiento permanente en Canarias  las etiquetas de impuestos les quedan perfectamente diferenciadas , no creo que les presente problema  te lo digo por experiencia que ya manejamos algunas con esa casuística

julianram committed:
[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

[ADD] l10n_es_atc_dua : Importaciones con DUAS para Canarias

etobella commented:
No digo de ponerlo en el l10n_es_igic, sino que el modulo que creais que tenga la dependencia de ambos para simplificar.

julianram commented:
@etobella , si  pones a depender l10n_es_dua_atc de l10n_es_dua instalamos dos modulos donde l10n_es_dua a las empresas Canarias no les aporta de nada, de todas forma si nos ilustras con un ejemplo ???

etobella commented:
Estais duplicando las posiciones fiscales y los impuestos DUA. De esta forma podeis reaprovecharlo todo

etobella commented:
A ver, me gustaría hacer un ejercicio de comparación entre ambos módulos. Empezando por el `l10n_es_dua`, se crean los siguientes registros:

- [Un grupo de Tasas llamado DUA Exento](https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_dua/data/tax_group_data.xml)
- [Una tasa llamada DUA Exento](https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_dua/data/taxes_dua.xml)
- [Una posición fiscal llamada Importación con DUA](https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_dua/data/fiscal_positions_dua.xml) con su consiguiente [mapeo para los IVAs de península](https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_dua/data/fiscal_position_taxes_dua.xml)
- [3 productos para la valoración del IVA y uno para la parte de compensación](https://github.com/OCA/l10n-spain/blob/16.0/l10n_es_dua/data/products_dua.xml)

Si miramos el módulo `l10n_es_atc_dua` podemos ver que los 3 primeros puntos son idénticos (de hecho, tienen hasta el mismo id que el de `l10n_es_dua`) pero con el mapeo del IGIC. En los productos has creado el de compensación igual y los productos para la valoración con sus respectivos IGIC.

Como puedes ver, estamos duplicando la misma configuración con la diferencia de los impuestos mapeados.

Que propongo, que este módulo solo añada los mapeos en la posición fiscal de `l10n_es_dua` y cree los productos para la valoración de IGIC. La única desventaja que habrá en este caso, es que los usuarios de canarias tendrán 3 productos más creados por el DUA con el IVA, pero podríamos archivarlos de forma automática al instalar el módulo (un post_install sería lo mejor para poder desarchivarlo si es necesario en los casos en los que un cliente tenga sede en canarias y en península). Las ventajas es que solo existirá una posición fiscal para DUA y, además, unificamos criterios y evitamos que haya diferencias según Canárias y península.

Quiero recalcar que esto es mi opinión, pero lo comento por que creo que facilitará la gestión y las desventajas de tener dos módulos me parecen ínfimas. En cualquier caso, si los PSC estan de acuerdo con duplicarlo todo yo no me voy a oponer.

@HaraldPanten

julianram commented:
@etobella  Partimos de la base de que no es necesario diferenciar las operaciones. Sin embargo, como mencionabas, especialmente para empresas acogidas a IVA e IGIC , cuanto más diferenciadas estén estas operaciones, mejor. Esto, aunque parezca que duplicamos no es correcto solo recogemos  la realidad de la ATC que es duplicar a la AEAT .

De todas formas no cuesta nada aplicar tu enfoque [17.0]  pero te adelanto que por aquí en Canarias lo que va a generar es mas confusión.

Como lo vean ...!!!

etobella commented:
En que sentido generará confusión? Lo digo para preparar a mis clientes canarios

HaraldPanten commented:
Tras darle varias vueltas, y a pesar de que la mejor idea es siempre reaprovechar todo el código posible, empiezo a dudar que hacer depender el módulo `l10n_es_atc_dua` de `l10n_es_dua` sea lo mejor.

Si nuestras compañías son solo de Canarias, con su plan contable, es cierto que no nos aportará nada el hecho de tener lo que nos aporta el módulo `l10n_es_dua`, correcto? Entonces si heredáramos `l10n_es_dua` tendríamos que:

- Archivar los productos para la valoración con IVA. Ya sea por medio del módulo, o el usuario manualmente.
- Podríamos reaprovechar el impuesto DUA Exento.
- Podríamos reaprovechar el producto DUA Compensación.
- Habría que mapear de distinta forma la posición fiscal de Importación con DUA, pero se podría reaprovechar la posición fiscal como tal, para entendernos.

¿Pero qué pasaría si tuviéramos multicompañía con multisede, una compañía en península y otra en Canarias? ¿En ese caso no sería más práctico tener los módulos por separado?

Ahora mismo, pienso que tenerlos por separado nos aporta más versatilidad, mientras que crear dependencias nos dificulta otras posibilidades, a pesar de beneficiarnos del "aprovechamiento de código".

**Módulos por separado:**

PROS:

- Simplificamos los casos con compañías Canarias, únicamente, ya que no se incorporan elementos innecesarios.
- Hacemos más sencillos los casos en que haya multilocalización (compañía canaria + compañía peninsular).
- Diferenciamos mejor cada una de las casuísticas.

CONTRAS:

- No reaprovechamos código.

**Módulos dependientes:**

PROS:

- Reaprovechamos código.

CONTRAS:

- Perdemos/complicamos la versatilidad de los mismos.

Otro punto a destacar... Pienso que el módulo debería llamarse `l10n_es_igic_dua`, ya que hereda del mismo y no de `l10n_es_atc`.

¿Cómo lo ves @etobella ? ¿Puede ser que no haya entendido del todo bien tu propuesta, o se me escape algo?

etobella commented:
Gracias por tus comentarios @HaraldPanten Me gustaria añadir un comentario a que al separar....

> Hacemos más sencillos los casos en que haya multilocalización (compañía canaria + compañía peninsular).

Creo que no es el caso. En el caso de multilocalización duplicamos productos e impuestos, complicandole la vida al usuario, ya que debe decidir cual es el impuesto y producto que debe seleccionar en su caso. Yo tengo un dicho, si le das dos opciones a un usuario, se equivocará...

Tengo que decir que es verdad que diferenciamos un poco mejor, però no tengo claro que sea necesario.

Por cierto, en que punto crees que perdemos versatilidad?

HaraldPanten commented:
Hola Enric, sí, te comento:

En los casos de multilocalización puedes limitar los productos para que estén disponibles en la compañía que corresponda. Los de DUA península limitados a península y los DUA IGIC limitados a Canarias. De este modo, el usuario no se confundirá, ya que directamente sólo podrá escoger los que sí o sí tendría que utilizar.

Diferenciarlo en dos módulos (sin herencias) implicará que sea más sencilla la gestión para las compañías que sólo tengan localización canaria. Simplemente instalarán el módulo y poco más.

Por otra parte, versatilidad me refería a que dos módulos (sin herencias) me parecen más sencillos que gestionar que uno con herencia de otro. Además, pienso que nos ofrecen más posibilidades, así como una gestión más sencilla cuando se trabaja con multilocalización (península + islas). Es verdad que los casos de multicompañía y multilocalización se darán un menor número de veces, pero ocurren.

Quizá @omar7r pueda aportar su granito de arena, pues esta cuestión creo que le impacta.

christian-ramos-tecnativa commented:
Me gustaría aportar un precedente: el modelo 415 de la ATC lo hicimos dependiente del 347 debido a la gran cantidad de código reutilizado, a pesar de que el usuario pueda no utilizarlo. Puedes ver el ejemplo aquí: [l10n_es_atc_mod415/manifest.py#L14](https://github.com/OCA/l10n-spain/blob/2c74ffdda5d9ffa9232fa517305fc5e9a1ca0343/l10n_es_atc_mod415/__manifest__.py#L14).

Creo que, como norma general, deberíamos enfocarnos en la reutilización de código, a menos que exista un caso muy claro que justifique lo contrario. Actualmente, al instalar `l10n_es_igic`, se instalan todos los impuestos de España, aunque muchas empresas en Canarias no los utilizan. Teniendo en cuenta que los usuarios tienen la capacidad de archivar estos impuestos/productos innecesarios, considero que la reutilización es una opción válida y eficiente.

julianram commented:
Insisto no es cuestión de reutilizar, que prácticamente es lo mismo, el l10n_es_atc_dua lo que aporta es claridad a la gestión diferenciando dos procesos que son claramente similares pero distintos antes las diferentes administraciones.

etobella commented:
@julianram Yo no le veo más claridad a copiarlo todo, al final no es questión de copiar solo, sinó que el mantenimiento sea más sencillo, y yo creo que tal y como te lo proponemos el módulo es mucho más sencillo y mantenible

HaraldPanten commented:
Una vez valoradas las opiniones, creo que lo que sugiere Enric es una propuesta mejor que lo que hablábamos anteriormente. @julianram Podrías trabajar la aportación en esa línea?

Muchas gracias de antemano.

julianram commented:
@HaraldPanten Sin problema, vamos por esa línea a ver si efectivamente es menos  problema,¿cerramos este PR?

etobella commented:
Podéis hacer push force así no se pierde la conversación.

HaraldPanten commented:
> @HaraldPanten Sin problema, vamos por esa línea a ver si efectivamente es menos problema,¿cerramos este PR?

Podéis hacerlo directamente en este PR, sin necesidad de abrir otro nuevo; como comenta Enric.

Saludos.

FernandoRomera commented:
Una propuesta, si al final heredáis de l10n_es_dua, podrías cambiar el nombre del módulo a l10n_es_dua_atc (igual que pasa con el l10n_es_dua_sii). De esta manera permanecerán agrupados en la lista.

HaraldPanten commented:
Sí, el nombre del módulo debería cambiarse. `l10n_es_igic_dua` o incluso `l10n_es_dua_igic`.

@etobella Qué opinas?

etobella commented:
`l10n_es_dua_igic` parece lo más límpio para mi

etobella commented:
Superseeded by #3958
